doc/settings.rst
Settings
--------

As a Django project, this project is configured via a file called :file:`settings.py`. You should
not edit this file yourself (it's in the git repository to provide defaults), but edit the
:file:`xmppaccount/localsettings.py` instead.  If you haven't already, copy the file
:file:`localsettings.py.example` in the same directory and start from there.

This page documents custom settings used by this project, for all other available settings, please
see `Django settings`_.

.. _Django settings: https://docs.djangoproject.com/en/dev/ref/settings/

.. _settings-BLOCKED_EMAIL_TLDS:

BLOCKED_EMAIL_TLDS
__________________

Default: ``set()`` (Empty set)

A set of domains that are not allowed to be used in Email addresses during registration or as a new
Email address for an existing user.

Domains from the :ref:`settings-XMPP_HOSTS` setting that don't have their ``EMAIL`` key set to
``True`` will be automatically be included in this list.

.. _settings-BRAND:

BRAND
_____

Default: ``''`` (Empty string)

The default "brand" displayed on the left of the navigation bar. If not set, the hostname of the
jabber-server (according to :ref:`settings-XMPP_HOSTS_MAPPING` is used instead.

CONTACT_URL
___________

Default:: ``''`` (Empty string)

The default URL where the user may contact the server administrators. This URL is only used when
:ref:`settings-XMPP_HOSTS` does not define a contact URL.

.. _settings-CLEARTEXT_PASSWORDS:

CLEARTEXT_PASSWORDS
___________________

Default:: ``True``

Unless this setting is set to ``False``, the site will display a warning that passwords are stored
in clear text whenever the user is prompted to enter a new password.

.. _settings-CONFIRMATION_TIMEOUT:

CONFIRMATION_TIMEOUT
____________________

Default:: ``timedelta(hours=24)`` (24 hours)

Every action requires an Email confirmation: The user receives an Email, and only if he clicks the
link provided there, he can actually do anything. This setting configures the time the links in the
comfirmation Emails stay valid.

.. _settings-DEFAULT_XMPP_HOST:

DEFAULT_XMPP_HOST
_________________

Default:: see text

The host that is preselected in the dropdown of the registration form. This setting of course has
an effect if you have more then one server. If you do not specify it yourself, an arbitrary host
will be selected on application startup.

.. _settings-EJABBERD_REGISTRATION_BOT:

EJABBERD_REGISTRATION_BOT
_________________________

Bot configuration for import of new registrations from ejabberd. See the
:ref:`ejabberd_registrations` management command for more information.

.. _settings-FORM_TIMEOUT:

FORM_TIMEOUT
____________

Default:: ``3600`` (1 hour)

Maximum amount of time a form stays valid. If the user loads a page and submits it after this many
seconds, he will have to reload the page to get a new form.

LOGDIR
______

Default:: ``./logs/`` (from the same directory as ``manage.py``)

Directory were some ``manage.py`` commands intended to be run as cron jobs will store their logs.
See :ref:`cron_jobs` for more information.

.. _settings-MIN_USERNAME_LENGTH:

MIN_USERNAME_LENGTH
___________________

Default:: ``3``

Minimum characters required for the registration of a new username.

.. _settings-MAX_USERNAME_LENGTH:

MAX_USERNAME_LENGTH
___________________

Default:: ``32``

Maximum characters allowed for the registration of a new username. If you provide a value greater
then 255 characters, it will be capped to that value.

.. _settings-RATELIMIT_WHITELIST:

RATELIMIT_WHITELIST
___________________

Default:: ``set()``

A list of IP addresses not affected by ratelimiting.

.. _settings-RECAPTCHA_PRIVATE_KEY:

RECAPTCHA_PRIVATE_KEY
_____________________

Default:: ``""`` (empty string)

If you set both this setting and :ref:`settings-RECAPTCHA_PUBLIC_KEY`, every form will be protected
by a CAPTCHA.

.. _settings-RECAPTCHA_PUBLIC_KEY:

RECAPTCHA_PUBLIC_KEY
____________________

Default:: ``""`` (empty string)

If you set both this setting and :ref:`settings-RECAPTCHA_PRIVATE_KEY`, every form will be
protected by a CAPTCHA.

.. _settings-REGISTRATION_RATE:

REGISTRATION_RATE
_________________

Default::

   {
       timedelta(minutes=2): 1,
       timedelta(hours=1): 2,
       timedelta(days=1): 5,
   }

This configures how many accounts a single IP address can register within the given timeframes.
Every restriction is added up, if any rate is exceeded, no registration is possible at the given
time. The default means that an IP address can register at most:

* once every two minutes
* twice every hour
* five times per day

If you want to override this setting, make sure you have the ``datetime.timedelta`` imported at the
top of :file:`xmpplist/localsettings.py`::

   from datetime import timedelta

.. _settings-SPAM_BLOCK_TIME:

SPAM_BLOCK_TIME
_______________

Default:: ``86400`` (24 hours)

If the client shows behaviour that clearly identifies it as spambot, it will be blocked for this
amount of seconds.

Clients are identified as spambots if:

* some hidden form fields are given incorrectly
* a form is submitted within one second

.. _settings-WELCOME_MESSAGE:

WELCOME_MESSAGE
_______________

Default:: ``None`` (No message will be sent).

If set to a dictionary, newly registered users will receive a welcome message via XMPP upon
registration. Example::

   WELCOME_MESSAGE = {
      'subject': "Welcome to {domain}!',
      'message': """Dear {username}@{domain},

   We hope you like our server. If you ever loose your password, go to
   {password_reset_url} and an email will be sent to {email}.

   yours, the {domain} team""",
   }

Standard formatting can be applied, as shown above. Availiabe keys are:

================== =====================================================
key
================== =====================================================
username           The username (excluding the domain).
domain             The domain the user registered at.
email              The email address he registered with.
contact_url        URL defined by the ``CONTACT_URL`` setting.
password_reset_url Full URL where the user can reset his/her password.
email_reset_url    Full URL where the user reset the email address.
delete_url         Full URL where the user can delete the account.
================== =====================================================

.. _settings-XMPP_BACKENDS:

XMPP_BACKENDS
_____________

Default:: ``{}`` (Empty dictionary, **required**)

Configure XMPP backends for this site. See :doc:`backends <backends>` for a list of available
backends and their settings. The only required setting is ``BACKEND``, which gives the Python path
to the implementation. Example::

   XMPP_BACKENDS = {
      'default': {
         'BACKEND': 'backends.ejabberd_xmlrpc.EjabberdXMLRPCBackend',
         'USER': 'account.example.com',
         'SERVER': 'account.example.com',
         'PASSWORD': 'super-secure-password',
      }
   }

Currently the only used key for this setting is ``default``.

.. _settings-XMPP_HOSTS:

XMPP_HOSTS
__________

Default: ``{}`` (Empty dictionary, **required**)

A dictionary of the hosts this installation is able to manage. This means that your backend (see
:ref:`settings-XMPP_BACKENDS`) can handle each of these domains.

.. WARNING:: Never remove hosts from this setting entirely unless you have (manually!) removed all
   users from this domain from the database.

   If you want this site to stop managing a given host, set its ``MANAGE`` setting to ``False``.

The value must be a dictionary, with the keys being domains and the values being dictionaries, with
the following possible keys:

.. rubric:: REGISTRATION

Default:: ``False``

A boolean ``True`` means that users should be able to register at this host.

.. rubric:: RESERVATION

Default:: ``False``

A boolean ``True`` means that the backend will reserve the username when the user first registers
(when the confirmation Email is sent to the user). For most backends this means that the user is
created with a randomly generated password and the real password is only set when the user provides
a password at the Email confirmation page.

.. rubric:: MANAGE

Default:: ``True``

Set this value to ``False`` if you want to completely disable a domain but still have local users
in the database.

.. rubric:: EMAIL

Default:: ``False``

Unfortunately people frequently try to give their full Jabber ID as their Email address. Unless you
set this setting to ``True``, users will not be able to fill in Email addresses with this domain in
any form.

.. rubric:: BRAND

Default:: :ref:`settings-BRAND` or :ref:`settings-DEFAULT_XMPP_HOST`

If set, display this brand instead. If you do not want to use a brand, set this to an empty string.

.. rubric:: CONTACT_URL

Default:: :ref:`settings-CONTACT_URL`

An URL where the user may contact the server administrators. This URL is displayed e.g. if the
registration fails, so the URL should provide at least some way to contact you **without** XMPP.

..rubric:: FROM_EMAIL

Default:: ``settings.DEFAULT_FROM_EMAIL``

The email address used in the ``From:`` field fo emails sent on behalf of this site.


Example::

   XMPP_HOSTS = {
      'jabber.at': {
         'REGISTRATION': True,
         'RESERVATION': True,  # users can also do in-band registration
      },
      'jabber.net': {
         'REGISTRATION': True,
         'RESERVATION': False,  # users can only register through this site
      }
      'oldhost.jabber.at': {
         'MANAGE': False,  # we used this many years back
      }
   }

.. _settings-XMPP_HOSTS_MAPPING:

XMPP_HOSTS_MAPPING
__________________

Default: ``{}``

A mapping of XMPP hosts to HTTP hosts this page is available under. This will change the behaviour
of your site depending on the hostname used. For example, if you use
``https://register.example.com`` to register accounts for ``example.com`` and
``https://account.example.org`` to register accounts for ``example.org``, use::

   XMPP_HOSTS_MAPPING = {
      'register.example.com': 'example.com',
      'account.example.org': 'example.org',
   }

The values of the dictionary must match a host defined in :ref:`settings-XMPP_HOSTS`. If the site
is viewed via an unknown domain (e.g. ``something-else.example.net`` in the above example, the
:ref:`settings-DEFAULT_XMPP_HOST` will be used.


xmppaccount/settings.py
# -*- coding: utf-8 -*-
#
# This file is part of django-xmpp-register (https://account.jabber.at/doc).
#
# django-xmpp-register is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# django-xmpp-register is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along with
# django-xmpp-register.  If not, see <http://www.gnu.org/licenses/>.

import gnupg
import logging
import os

from datetime import timedelta
log = logging.getLogger(__name__)

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(__file__))

DEBUG = False
TEMPLATE_DEBUG = DEBUG
RECAPTCHA_CLIENT = None

ADMINS = (
    # ('Your Name', 'your_email@example.com'),
)

MANAGERS = ADMINS

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3', # Add 'postgresql_psycopg2', 'mysql', 'sqlite3' or 'oracle'.
        'NAME': '.xmppaccount.sqlite3',                      # Or path to database file if using sqlite3.
        # The following settings are not used with sqlite3:
        'USER': '',
        'PASSWORD': '',
        'HOST': '',                      # Empty for localhost through domain sockets or '127.0.0.1' for localhost through TCP.
        'PORT': '',                      # Set to empty string for default.
    }
}

# Hosts/domain names that are valid for this site; required if DEBUG is False
# See https://docs.djangoproject.com/en/1.5/ref/settings/#allowed-hosts
ALLOWED_HOSTS = []

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# In a Windows environment this must be set to your system time zone.
TIME_ZONE = 'Europe/Vienna'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'en-us'

# If you set this to False, Django will make some optimizations so as not
# to load the internationalization machinery.
USE_I18N = True

# If you set this to False, Django will not format dates, numbers and
# calendars according to the current locale.
USE_L10N = True

# If you set this to False, Django will not use timezone-aware datetimes.
USE_TZ = True

# Absolute filesystem path to the directory that will hold user-uploaded files.
# Example: "/var/www/example.com/media/"
MEDIA_ROOT = ''

# URL that handles the media served from MEDIA_ROOT. Make sure to use a
# trailing slash.
# Examples: "http://example.com/media/", "http://media.example.com/"
MEDIA_URL = ''

# Absolute path to the directory static files should be collected to.
# Don't put anything in this directory yourself; store your static files
# in apps' "static/" subdirectories and in STATICFILES_DIRS.
# Example: "/var/www/example.com/static/"
STATIC_ROOT = ''

# URL prefix for static files.
# Example: "http://example.com/static/", "http://static.example.com/"
STATIC_URL = '/static/'

# Additional locations of static files
STATICFILES_DIRS = (
    # Put strings here, like "/home/html/static" or "C:/www/django/static".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
)

# List of finder classes that know how to find static files in
# various locations.
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
#    'django.contrib.staticfiles.finders.DefaultStorageFinder',
)

# Make this unique, and don't share it with anybody.
SECRET_KEY = 'igw!e%gbu_u+e%+$*d3u_(lo#ixs2pqrv@9zzzw(mpkx6q9qm+'

# add request preprocessor to default
TEMPLATE_CONTEXT_PROCESSORS = (
    "django.contrib.auth.context_processors.auth",
    "django.core.context_processors.debug",
    "django.core.context_processors.i18n",
    "django.core.context_processors.media",
    "django.core.context_processors.request",
    "django.core.context_processors.static",
    "django.core.context_processors.tz",
    "django.contrib.messages.context_processors.messages",
    "core.context_processors.xmppaccount",
)

MIDDLEWARE_CLASSES = (
    'core.middleware.AntiSpamMiddleware',
    'core.middleware.SiteMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    # Uncomment the next line for simple clickjacking protection:
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
)

ROOT_URLCONF = 'xmppaccount.urls'

# always upload to temporary files right away, since gnupg needs files there
FILE_UPLOAD_HANDLERS = ("django.core.files.uploadhandler.TemporaryFileUploadHandler",)

# Python dotted path to the WSGI application used by Django's runserver.
WSGI_APPLICATION = 'xmppaccount.wsgi.application'

TEMPLATE_DIRS = (
    # Put strings here, like "/home/html/django_templates" or "C:/www/django/templates".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
)

INSTALLED_APPS = (
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    # Uncomment the next line to enable the admin:
    'django.contrib.admin',
    # Uncomment the next line to enable admin documentation:
    # 'django.contrib.admindocs',

    'core',
    'register',
    'reset',
    'delete',
)

SESSION_SERIALIZER = 'django.contrib.sessions.serializers.JSONSerializer'

AUTHENTICATION_BACKENDS = (
    'core.auth_backends.BackendBackend',
)

AUTH_USER_MODEL = 'core.RegistrationUser'

# custom settings defaults
XMPP_HOSTS = {}
XMPP_HOSTS_MAPPING = {}
DEFAULT_XMPP_HOST = None
MIN_USERNAME_LENGTH = 3
MAX_USERNAME_LENGTH = 32

CLEARTEXT_PASSWORDS = True
CONFIRMATION_TIMEOUT = timedelta(hours=24)
FORM_TIMEOUT = 60 * 60  # 1 hour

SPAM_BLOCK_TIME = 60 * 60 * 24  # one day!
REGISTRATION_RATE = {
    timedelta(minutes=2): 1,
    timedelta(hours=1): 2,
    timedelta(days=1): 5,
}
BLOCKED_EMAIL_TLDS = set()

RECAPTCHA_PRIVATE_KEY = None
RECAPTCHA_PUBLIC_KEY = None
BRAND = ""
CONTACT_URL = ""
WELCOME_MESSAGE = None

# logging
LOGDIR = os.path.join(BASE_DIR, 'logs')
LOG_LEVEL = 'WARNING'
RATELIMIT_WHITELIST = set()

# GPG config
GNUPG = {
    'gnupghome': os.path.join(BASE_DIR, '.gpg'),
    'gpgbinary': '/usr/bin/gpg',
}
FORCE_GPG_SIGNING = False

try:
    from xmppaccount.localsettings import *
except ImportError as e:
    print(e)
    pass

MANAGED_HOSTS = [k for k, v in XMPP_HOSTS.items() if v.get('MANAGE', True)]
RESERVATION_HOSTS = [k for k, v in XMPP_HOSTS.items()
                     if v.get('RESERVATION') and v.get('MANAGE', True)]
REGISTRATION_HOSTS = [k for k, v in XMPP_HOSTS.items()
                      if v.get('REGISTRATION') and v.get('MANAGE', True)]
NO_EMAIL_HOSTS = [k for k, v in XMPP_HOSTS.items() if not v.get('EMAIL')]
BLOCKED_EMAIL_TLDS.update(NO_EMAIL_HOSTS)

if MAX_USERNAME_LENGTH > 255:
    MAX_USERNAME_LENGTH = 255

if DEFAULT_XMPP_HOST is None:
    DEFAULT_XMPP_HOST = list(XMPP_HOSTS.values())[0]

if RECAPTCHA_PRIVATE_KEY and RECAPTCHA_PUBLIC_KEY:
    from recaptcha import RecaptchaClient
    RECAPTCHA_CLIENT = RecaptchaClient(RECAPTCHA_PRIVATE_KEY, RECAPTCHA_PUBLIC_KEY)


GPG = None
if GNUPG is not None:
    if os.path.exists(GNUPG['gnupghome']):
        GPG = gnupg.GPG(**GNUPG)
    else:
        log.warn('GnuPG disabled because GnuPG home not found. Generate key with manage.py genkey.')

if not os.path.exists(LOGDIR):
    os.makedirs(LOGDIR)

# A sample logging configuration. The only tangible logging
# performed by this configuration is to send an email to
# the site admins on every HTTP 500 error when DEBUG=False.
# See http://docs.djangoproject.com/en/dev/topics/logging for
# more details on how to customize your logging configuration.
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'filters': {
        'require_debug_false': {
            '()': 'django.utils.log.RequireDebugFalse',
        },
    },
    'formatters': {
        'simple': {
            'format': '%(levelname)-8s %(message)s',
        },
        'cron': {
            'format': '[%(asctime).19s] %(message)s',  # .19s = only first 19 chars
        },
    },
    'handlers': {
        'mail_admins': {
            'level': 'ERROR',
            'filters': ['require_debug_false'],
            'class': 'django.utils.log.AdminEmailHandler',
        },
        'console': {
            'level': LOG_LEVEL,
            'class': 'logging.StreamHandler',
            'formatter': 'simple',
        },
        'cron-console': {  # always log error to stdout
            'level': 'ERROR',
            'class': 'logging.StreamHandler',
            'formatter': 'cron',
        },
        'cleanup': {
            'level': 'INFO',
            'class': 'logging.FileHandler',
            'filename': os.path.join(LOGDIR, 'cleanup.log'),
            'formatter': 'cron',
        },
        'import': {
            'level': 'INFO',
            'class': 'logging.FileHandler',
            'filename': os.path.join(LOGDIR, 'import.log'),
            'formatter': 'cron',
        },
    },
    'loggers': {
        'django.request': {
            'handlers': ['mail_admins'],
            'level': 'ERROR',
            'propagate': True,
        },
        'backends': {
            'handlers': ['console', ],
            'level': LOG_LEVEL,
        },
        'core': {
            'handlers': ['console', ],
            'level': LOG_LEVEL,
        },
        'cleanup': {
            'handlers': ['cleanup', 'cron-console'],
            'level': LOG_LEVEL,
        },
        'import': {
            'handlers': ['import', 'cron-console'],
            'level': LOG_LEVEL,
        },
    },
}
